File: decode.cpp
#include <cstddef>
#include <cstdint>
#include <iomanip>
#include <iostream>
#include <tuple>
#include <vector>

/**
 * @brief Decode TAXI timestamp signal
 *
 * @param trace oscilloscope trace
 * @param bit_length oscilloscope trace
 * @param num_bits number of bits in a frame
 * @param threshold threshold for high / low
 *
 * @return list of successfully decoded frames, as tuples of <sample_index,
 * frame_data>
 */
std::vector<std::tuple<int, uint64_t>> decode_taxi_time(
    std::vector<uint16_t> trace, double bit_length, size_t num_bits,
    uint16_t threshold) {
    const double frame_gap = bit_length * 9.5;  // gap between frames
    size_t index_start = 0;                     // start index of current frame
    std::vector<std::tuple<int, uint64_t>> frames = {};
    while (true) {
        // find interframe gap (no edges for > frame_gap)
        size_t last_edge_index = index_start;
        bool state_last = false;
        // loop through array
        for (size_t i = index_start; i < trace.size(); i++) {
            bool state = trace[i] > threshold;
            // edge ?
            if (state == state_last) {
                continue;
            }
            state_last = state;
            // gap between edges long enough?
            if ((i - last_edge_index) < frame_gap) {
                last_edge_index = i;
                continue;
            }
            // frame found, go to decoding step
            index_start = i;
            goto frame_start;
        }
        break;  // loop exited normally -> reached end of array
    frame_start:
        // frame found, decode frame
        uint64_t code = 0;
        for (size_t i_bit = 0; i_bit < num_bits; i_bit++) {
            // skip synchronization bits (every eight bit)
            size_t i_bit_eff = i_bit + i_bit / 8;
            // sample frame between edges, starting at the first data bit
            double index_offset =
                (static_cast<double>(i_bit_eff) + 1.5) * bit_length;
            size_t i_sample = index_start + static_cast<size_t>(index_offset);
            // check bounds
            if (i_sample > trace.size()) {
                goto end_of_trace;
            }
            // sample
            code <<= 1;
            if (trace[i_sample] > threshold) {
                code |= 1;
            }
        }
        frames.push_back({index_start, code});
    }
end_of_trace:
    return frames;
}

// only for testing
#include <fstream>
int main(int argc, char **argv) {
    double interval = 1.6000000213622911e-09;
    double bit_length = 1 / 25e6 / interval;
    std::vector<uint16_t> data;
    // load data from file
    std::ifstream infile("test.txt");
    uint16_t value;
    while (infile >> value) {
        data.push_back(value);
    }
    // decode
    auto frames = decode_taxi_time(data, bit_length, 35, 128);
    // print
    for (auto frame : frames) {
        size_t index = std::get<0>(frame);
        uint64_t code = std::get<1>(frame);

        uint64_t whole = code >> 19;
        uint64_t fractional = code & ((1 << 19) - 1);
        double time =
            static_cast<double>(whole) + static_cast<double>(fractional) * 2e-6;
        std::cout << index << " " << std::fixed << std::setprecision(6)
                    << time << std::endl;
    }
}

File: decode.py
#!/usr/bin/python
import scipy.io
import numpy as np
from pathlib import Path
import matplotlib.pyplot as plt

def decode_channel(chan: np.ndarray, interval: float, baud: float = 25e6, num_bits: int = 35):
    # timing
    bit = 1 / baud / interval # 25 MHz baud rate
    frame_gap = 8.5 * bit
    # threshold and edge detection
    chan = chan > 0
    chan_edge = np.logical_xor(chan, np.roll(chan, 1))
    # decoder state
    frames = []
    # go through data
    index_start = 0
    while True:
        last_edge_index = index_start
        for i in range(index_start, len(chan)):
            if not chan_edge[i]:
                continue
            if (i - last_edge_index) < frame_gap:
                last_edge_index = i
                continue
            index_start = i
            break
        else:
            break
        code = 0
        samples = []
        for i_bit in range(num_bits):
            i_bit_eff = i_bit + (i_bit // 8)
            i_bit_eff = index_start + int((i_bit_eff + 1.5) * bit)
            if i_bit_eff >= len(chan):
                break # prevents outer else from running -> breaks outer loop
            samples.append(i_bit_eff)
            code |= chan[i_bit_eff] << (num_bits - 1 - i_bit)
        else:
            sec_of_day = code >> 19
            sub_second = code & ((1 << 19) - 1)
            frame_time = sec_of_day + sub_second * 2e-6
            frames.append((index_start, frame_time))
            # print(f'sec: 0x{sec_of_day:X} subsec: 0x{sub_second:X}')
            # print(f'frame at {index_start}: {frame_time:0.6f}')
            continue
        break
    return frames

last_time = 0
def decode_mat(p: Path) -> tuple[np.ndarray, float]:
    global last_time
    mat = scipy.io.loadmat(p)
    # channels = []
    interval = mat['Tinterval'][0, 0]
    for name in ('A', 'B', 'C', 'D'):
        if name not in mat:
            continue
        frames = decode_channel(mat[name][:,0], interval)
        if len(frames) == 0:
            print('NO FRAMES', last_time)
            continue
        last_time = frames[0][1]

for i in range(512):
    # decode_mat(f'data/triggered/triggered_{i+1:03}.mat')
    decode_mat(f'data/untriggered/untriggered_{i+1:03}.mat')
    # decode_mat(f'data/untriggered_rapid/untriggered_rapid_{i+1:03}.mat')


mat = scipy.io.loadmat('data/triggered/triggered_001.mat')
interval = mat['Tinterval'][0, 0]
data = mat['A'][:,0]

print('interval', interval)
with open('test.txt', 'w') as f:
    for x in data:
        print(int(255 * (x + 1)/2), file=f)
